src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -10,11 +10,54 @@
 #define FOH_COMMIT_SIZE (64 * 1024)
 
 FrozenObjectHeapManager::FrozenObjectHeapManager():
-    m_Crst(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC),
+    // This lock is used in both COOP and PREEMP (by profiler) modes
+    m_Crst(CrstFrozenObjectHeap,CRST_UNSAFE_ANYMODE),
+    // This lock is used only in COOP mode
+    m_InitSegmentCrst(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC),
     m_CurrentSegment(nullptr)
 {
 }
 
+void FrozenObjectHeapManager::CreateNewSegment(CrstHolderWithState* activeLock)
+{
+    size_t size = FOH_SEGMENT_DEFAULT_SIZE;
+    if (m_CurrentSegment != nullptr)
+    {
+        // Double the reserved size to reduce the number of frozen segments in apps with lots of frozen objects
+        // Use the same size in case if prevSegmentSize*2 operation overflows.
+        const size_t prevSegmentSize = m_CurrentSegment->GetSize();
+        size = max(prevSegmentSize, prevSegmentSize * 2);
+    }
+
+    const auto segment = new FrozenObjectSegment(size);
+
+    // Release the active lock since segment's initialization logic (RegisterFrozenSegment to be exact) might
+    // hit a gc lock while e.g. profiler is trying to enumerate NonGC objects in GC's callback and waiting for this lock.
+    activeLock->Release();
+
+    bool justInitialized = false;
+    {
+        // We don't want to initialize it twice so a new quick lock is taken.
+        CrstHolder initSegmentLock(&m_InitSegmentCrst);
+
+        // Double-checked locking if some thread already initialized it here.
```

Comment:
The new segment is not visible to other threads at this point. This won't ever happen in this implementation.

Reply:
Ah, I take the registration lock earlier than I release the main one - I pushed a slightly wrong version